### PR TITLE
Eliminate `no_grads` and squares in double backward tests

### LINKED ISSUE #4656: Careless use of `no_grads` in tests
The argument `no_grads` appears in the tests of
- `F.crf1d`
- `F.minimum`
- `F.sign`

This skips computing numerical gradients w.r.t. the no_grads inputs, while it seems authors just wanted "trivial" tests to pass.  I suggest `gx == None` should be checked as if `gx == xp.zeros_like(x)`, since such authors' math is not so "trivial" to everyone.

An exception is the test of `F.roi_pooling_2d` where `no_grads` is set to `rois` probably because indices are stored at `rois[:, 0]`:
```
            idx, xmin, ymin, xmax, ymax = bottom_rois[i_roi]
            xmin = int(round(xmin * self.spatial_scale))
            xmax = int(round(xmax * self.spatial_scale))
            ymin = int(round(ymin * self.spatial_scale))
            ymax = int(round(ymax * self.spatial_scale))
```

BTW, the tests of `L.SimplifiedDropconnect` specify default `no_grads`. I found removing them after the commit 56ed844 is OK:
```
% git diff
diff --git a/tests/chainer_tests/links_tests/activation_tests/test_simplified_dropconnect.py b/tests/chainer_tests/links_tests/activation_tests/test_simplified_dropconnect.py
index a7cf6a18c..564b442df 100644
--- a/tests/chainer_tests/links_tests/activation_tests/test_simplified_dropconnect.py
+++ b/tests/chainer_tests/links_tests/activation_tests/test_simplified_dropconnect.py
@@ -86,7 +86,7 @@ class TestSimplifiedDropconnect(unittest.TestCase):
         gradient_check.check_backward(
             self.link_wrapper, (x_data, mask), y_grad,
             (self.link.W, self.link.b), eps=2 ** -3,
-            no_grads=(False, True), **self.check_backward_options)
+             **self.check_backward_options)

     @condition.retry(3)
     def test_backward_cpu(self):
@@ -153,7 +153,7 @@ class TestSimplifiedDropconnectParameterShapePlaceholder(unittest.TestCase):
     def check_backward(self, x_data, y_grad, mask):
         gradient_check.check_backward(
             self.link_wrapper, (x_data, mask), y_grad,
-            (self.link.W, self.link.b), eps=1e-2, no_grads=(False, True))
+            (self.link.W, self.link.b), eps=1e-2, )
% nosetests -a '!gpu' chainer_tests/links_tests/activation_tests/test_simplified_dropconnect.py
........................................
----------------------------------------------------------------------
Ran 40 tests in 0.286s

OK
```

### PR DESCRIPTION
**Pull request by toslunar**
Use #4654 and stop squaring outputs of affine functions.

This PR also fixes #4656.

### CODE REVIEW

**Comment by toslunar** [comment]
This PR may improve accuracy of numerical/backprop gradients in the tests with `check_double_backward`.

File: tests/chainer_tests/functions_tests/math_tests/test_sign.py [hunk 1]
```diff
@@ -20,7 +20,9 @@ def setUp(self):
         self.x = numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
         self.gy = numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
         self.ggx = numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
-        self.no_grads = (True,)  # Always assert that gradients are `None`
+
+        # Avoid non-differentiable point
+        self.x[(abs(self.x) < 1e-2)] = 1
```
[the remark this answers is not in the document]
Comment (kmaehashi) [reply tests/chainer_tests/functions_tests/math_tests/test_sign.py]: Thanks, I see.

File: tests/chainer_tests/functions_tests/math_tests/test_sign.py [hunk 2]
```diff
@@ -53,22 +55,22 @@ def test_forward_ndarray_cpu(self):
     def test_forward_ndarray_gpu(self):
         self.check_forward_ndarray(cuda.to_gpu(self.x), cuda.cupy)
 
-    def check_backward(self, x_data, y_grad, no_grads):
+    def check_backward(self, x_data, y_grad):
         gradient_check.check_backward(
-            F.sign, x_data, y_grad, no_grads=no_grads)
+            F.sign, x_data, y_grad)
```
[the remark this answers is not in the document]
Comment (toslunar) [reply tests/chainer_tests/functions_tests/math_tests/test_sign.py]: Yes. I updated the commented of the PR accordingly.

[the remark this answers is not in the document]
Comment (toslunar) [reply tests/chainer_tests/functions_tests/math_tests/test_sign.py]: Yes. I updated the comment of the PR accordingly.

**Comment by toslunar** [comment]
I reverted deletion in `unary_math_function_test`. The squaring behavior with `is_linear=True` is kept as it was documented.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
LGTM!

**Pull request merged**